sound/soc/sof/ipc4-loader.c

```diff
@@ -178,6 +282,30 @@ struct sof_ipc4_fw_module *sof_ipc4_find_module_by_uuid(struct snd_sof_dev *sdev
 		}
 	}
 
+	/*
+	 * Do not attempt to load external library in case the maximum number of
+	 * firmware libraries have been already loaded
+	 */
+	if ((lib_id + 1) == ipc4_data->max_libs_count) {
+		dev_err(sdev->dev,
+			"%s: Maximum allowed number of libraries reached (%u)\n",
+			__func__, ipc4_data->max_libs_count);
+		return NULL;
+	}
+
+	/* The module can not be found, try to load it as a library */
+	ret = sof_ipc4_load_library_by_uuid(sdev, lib_id + 1, uuid);
+	if (ret)
+		return NULL;
+
+	/* Look for the module in the newly loaded library, it should be available now */
+	xa_for_each_start(&ipc4_data->fw_lib_xa, lib_id, fw_lib, lib_id) {
+		for (i = 0; i < fw_lib->num_modules; i++) {
+			if (guid_equal(uuid, &fw_lib->modules[i].man4_module_entry.uuid))
+				return &fw_lib->modules[i];
+		}
+	}
```

Comment:
can we condense the function to do this search only once by calling sof_ipc4_load_library_by_uuid() unconditionally first ? If the library is already loaded, this function should pretty much do nothing

Reply:
But then we should have similar check in the load library function to check if the requested library has been already loaded.
The first loop will check the loaded libraries, if the module is found it returns. If not then loads a library from the filesystem, parses the ext_manifest, then we will start to look for the new module starting _after_ the last library we have checked in the first loop, so we will continue the loop with the just loaded lib.

I don't want to put extra logic to load llibrary to deal with finding and matching modules, it is not it's job.